docs/design/auditing-containment.md

```diff
@@ -0,0 +1,168 @@
+
+By "uncooperative," I mean that the auditor will be able to successfully DialNode (within pkg/audit/verifier.go), but errors will occur at
+```
+downloader, err := ps.Download(timedCtx, limit.GetLimit(), offset, int64(shareSize))
+```
+or
+```
+_, err = io.ReadFull(downloader, buf)
+```
+Once these errors have occurred and we know that a `pendingAudit` entry should be made in the ContainmentDB, we should use infectious’s `f.Decode` to generate the original stripe.
+Then we should use infectious’s `f.EncodeSingle` where we input the stripe and output the missing share. We then make a hash of that missing share to be saved to the ContainmentDB as `ExpectedShareHash`.
+
+`containment.Contain` should be called when a node is found to be “uncooperative."
+
+``` pkg/audit/verifier.go
+verifier.containment.Contain(ctx, &pendingAudit{
+    nodeID:            nodeID,
+    pieceID:           pieceID,
+    pieceNum:          pieceNum,
+    stripeIndex:       stripeIndex,
+    shareSize:         shareSize,
+    expectedShareHash: shareHash,
+}
+```
+
+#### 2. Audits continue normally. But now the nodes listed in a pointer are checked for `contained` status.
+``` pkg/audit/containment/checker.go
+func (verifier *Verifier) Verify(ctx context.Context, stripe *Stripe) (verifiedNodes *RecordAuditsInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	pointer := stripe.Segment
+
+	pieces := pointer.GetRemote().GetRemotePieces()
+	
+	// get the NodeDossiers from the overlay using the pieces’ nodeIDs
+  // determine if they have the contained flag set to true
+  ...
+```
+
+#### 3. Contained nodes are checked for the same data that they were originally requested to respond with:
+
+```
+func (verifier *Verifier) Reverify(ctx context.Context, node storj.NodeID) (auditRecords *RecordAuditsInfo, err error) {
+
+  pendingAudit, err := checker.containment.Get(id)
+
+  // dial the node, then use info from the pendingAudit to download the target share from the node
+
+  offset := node.stripeIndex * node.shareSize
+  downloader, err := ps.Download(timedCtx, limit.GetLimit(), offset, int64(shareSize))
+
+  // if the error is a timeout (or any other error?), call ReverifyFail
+  // if the download occurred successfully, then get the hash of the original erasure share from the ContainmentDB
+
+  shareData := make([]byte, shareSize)
+  _, err = io.ReadFull(downloader, shareData)
+
+  // create a hash of shareData
+
+  var successNodes storj.NodeIDList
+  var failNodes storj.NodeIDList
+
+  if !bytes.Equal(shareHash, pendingAudit.expectedShareHash) {
+    failNodes = append(failNodes, node)
+    auditRecords.FailNodeIDs = failNodes
+  } else {
+    successNodes = append(successNodes, node)
+    auditRecords.SuccessNodeIDs = successNodes
+  }
+
+  // remove the set the contained flag to false on the NodeDossier and update the overlay (satellitedb)
+  // remove the pending audit from the ContainmentDB
+
+  return auditRecords
+```
+
+## Open Issues
```

Comment:
Can a storage node determine if they're in containment mode?  If so, can they do anything about it?  A storage node operator would want there to be a function where I can manually fix my containment mode status.

Reply:
I don't think a storage node should be able to determine if they're in containment mode. It would be a field on the NodeDossier. @kaloyan-raev do you know if there's a way that a SNO could manipulate their NodeDossier to say they're not in containment mode?

Reply:
Why?  Clever people can certainly just read the source code and reason.  Its the not-so-clever people I'm worried about.

Reply:
seems to maybe be out of scope for now